csrc/torch_memory_saver.cpp

```diff
@@ -108,42 +109,72 @@ namespace CUDAUtils {
         accessDesc.location.type = CU_MEM_LOCATION_TYPE_DEVICE;
         accessDesc.location.id = device;
         accessDesc.flags = CU_MEM_ACCESS_FLAGS_PROT_READWRITE;
+
         CURESULT_CHECK(cuMemSetAccess((CUdeviceptr) ptr, size, &accessDesc, 1));
     }
 }
 
 // ----------------------------------------------- primary class --------------------------------------------------
 
+class TmsThreadLocalContext {
+public:
+    TmsThreadLocalContext() : is_interesting_region_(false), tag_("default") {}
+    
+    bool is_interesting_region_;
+    std::string tag_;
+};
+
+static thread_local TmsThreadLocalContext current_thread_context_;
+
 struct _AllocationMetadata {
     size_t size;
     CUdevice device;
     CUmemGenericAllocationHandle allocHandle;
+    std::string tag;  // Add tag for memory type classification
 };
 
 class TorchMemorySaver {
 public:
-    TorchMemorySaver() {}
+    static TorchMemorySaver& getInstance() {
+        static TorchMemorySaver instance;
+        return instance;
+    }
+
+    // Delete copy constructor and assignment operator
+    TorchMemorySaver(const TorchMemorySaver&) = delete;
+    TorchMemorySaver& operator=(const TorchMemorySaver&) = delete;
+
+    void set_current_tag(const std::string& tag) {
+        current_thread_context_.tag_ = tag;
+    }
+
+    bool is_interesting_region() const {
+        return current_thread_context_.is_interesting_region_;
+    }
+
+    void set_interesting_region(bool enabled) {
+        current_thread_context_.is_interesting_region_ = enabled;
+    }
```

Comment:
I personally feel it good to make these in a separate class/namespace, for example originally I used a `namespace RegionManager` to separate the concerns